**WARNING: miniprot is WIP and not ready for production use.**

## Getting Started
```sh
git clone https://github.com/lh3/miniprot
cd miniprot && make
./miniprot test/DPP3-hs.gen.fa.gz test/DPP3-mm.pep.fa.gz
man ./miniprot.1  # for output format
```

## Introduction

Miniprot aligns a protein sequence against a genome with splicing and
frameshift. It is primarily intended for annotating protein-coding genes in a new
species using known genes from other species. Miniprot is similar to GeneWise
in functionality but it can map proteins to whole genomes and is over 100 times
faster at the residue alignment step.

Miniprot is not optimized for mapping distant homologs because distant homologs
are less informative to gene annotations. Nonetheless, you can tune seeding
parameters to achieve higher sensitivity at the cost of performance.

## Users' Guide

### Installation

Miniprot requires SSE2 or NEON instructions and only works on x86\_64 or ARM
CPUs. It depends on zlib for parsing gzip'd input files. To compile miniprot,
type `make` in the source code directory. This will produce a standalone
executable `miniprot`. This executable is all you need to invoke miniprot.

### Usage

To run miniprot, use
```sh
miniprot -t8 ref-file protein.faa > output.paf
```
where `ref-file` can either be a genome in the FASTA format or a pre-built
index generated by
```sh
miniprot -t8 -d ref.mpi ref.fna
```
Because miniprot indexing is slow and memory intensive, it is recommended to
pre-build the index. FASTA input files can be optionally compressed with gzip.

Miniprot outputs alignment in the protein PAF format. Different from the more
common nucleotide PAF format, miniprot uses more CIGAR operators to encode
introns and frameshifts. Please refer to the manpage for detailed explanation.

## Limitations

* Miniprot is not well tested and not well evaluated.

* The initial conditions of dynamic programming are not properly set which may
  result in suboptimal residue alignment in rare cases.
